# HOMINID
Python MPI program to identify associations between host genetic variation 
and microbiome taxonomic composition.

**Summary**: Recent studies have uncovered a strong effect of host genetic variation on the  composition of host-associated microbiota. Here, we present HOMINID, a computational approach based on Lasso linear regression, that given host genetic variation and microbiome composition data, identifies host SNPs that are correlated with microbial taxa abundances. By using HOMINID on data from the Human Microbiome Project, we identified 2,158 human SNPs in which genetic variation is correlated with microbiome taxonomic composition in 15 body sites. We also present a tool for visualization of host-microbiome association network identified in HOMINID.

Online visualization tool at http://z.umn.edu/genemicrobe

Contact: blekhmanlab@gmail.com

# Running the HOMINID software

You'll want to read these documents, in order:

1. [Python software requirements](https://github.com/blekhmanlab/hominid/wiki/Requirements)
2. [Setting up the Python virtual environment](https://github.com/blekhmanlab/hominid/wiki/Set-up-a-virtualenv-on-Linux)
3. [Installation](https://github.com/blekhmanlab/hominid/wiki/Installation)
4. [HOMINID analysis pipeline](https://github.com/blekhmanlab/hominid/wiki/HOMINID-pipeline)
5. [Running `hominid` on sample data](https://github.com/blekhmanlab/hominid/wiki/Running-hominid-on-sample-data)
6. [Running `hominid` on your data](https://github.com/blekhmanlab/hominid/wiki/Running-hominid-on-your-data)
7. [Running `hominid_stability_selection` on your data](https://github.com/blekhmanlab/hominid/wiki/Running-stability-selection-on-your-data)
8. [Running `hominid_sort_results` on your data](https://github.com/blekhmanlab/hominid/wiki/Running-sort-results-on-your-data)
